### ISSUE TITLE
Pathfinding: replace bfs with bidirectional search

### ISSUE THREAD

**Issue by Annoraaq** [Annoraaq/grid-engine#85]
Replace BFS algorithm with (faster) bidirectional search.

**Comment by Annoraaq** [comment]
Replace BFS algorithm with (faster) bidirectional search.

**Comment by borkowsky** [comment]
Where it comes? :)

**Comment by Annoraaq** [comment]
I am currently extending the movement to support 8 directions.

Are you currently facing problems with the performance of pathfinding?

If so I could give it a higher priority.

**Comment by borkowsky** [comment]
> I am currently extending the movement to support 8 directions.
> 
> Are you currently facing problems with the performance of pathfinding?
> 
> If so I could give it a higher priority.

Its a good news) yes, sometines characters stucks at path finding

**Comment by Annoraaq** [comment]
How large is your map? Bidirectional search will allow twice as long paths to be found in the same time than BFS.

**Comment by borkowsky** [comment]
> How large is your map? Bidirectional search will allow twice as long paths to be found in the same time than BFS.

50x50 tiles (10x10)

**Comment by Annoraaq** [comment]
This should not be a problem for BFS 🤔 

Maybe the following factors influence it as well:
1. how many chars do you have on the map performing pathfinding?
2. how many layers does your map have?
3. are you using the [Height Shift feature](https://annoraaq.github.io/grid-engine/features/height-shift.html)?

**Comment by borkowsky** [comment]
> This should not be a problem for BFS 🤔 
> 
> The following factors can influence it as well:
> 1. how many chars do you have on the map performing pathfinding?
> 2. how many layers does your map have?
> 3. are you using the [Height Shift feature](https://annoraaq.github.io/grid-engine/features/height-shift.html)?

Its happens when some another characters near my character.
1. 5-6
2. 5
3. No.

**Comment by Annoraaq** [comment]
Then it is not the speed of the algorithm.

Is it just "lagging" or does the character change directions a couple of times?

On every step the character is looking for the shortest path to the target again. If the other character (close to yours) is blocking that path, a new path is chosen. That might explain changes of direction.

**Comment by borkowsky** [comment]
> Then it is not the speed of the algorithm.
> 
> Is it just "lagging" or does the character change directions a couple of times?
> 
> On every step the character is looking for the shortest path to the target again. If the other character (close to yours) is blocking that path, a new path is chosen. That might explain changes of direction.

Yes, char change directions couple of times

**Comment by Annoraaq** [comment]
Then I guess the problem is created in a situation like this:

1.  char c1 wants to move to target X and calculates shortest path
2. char c2 gets in that path
3. char c1 needs to calculate a new shortest path, because the old one is blocked by c2
and so on

I am not sure how that could be prevented. One alternative is that c1 would just ignore c2 and continues walking on that path. But it will block until c2 will get out of the way.

Do you have a suggestion for a desired behavior?

**Comment by borkowsky** [comment]
> Then I guess the problem is created in a situation like this:
> 
> 1.  char c1 wants to move to target X and calculates shortest path
> 2. char c2 gets in that path
> 3. char c1 needs to calculate a new shortest path, because the old one is blocked by c2
> and so on
> 
> I am not sure how that could be prevented. One alternative is that c1 would just ignore c2 and continues walking on that path. But it will block until c2 will get out of the way.
> 
> Do you have a suggestion for a desired behavior?
> 
> Edit: Maybe the effect could be weakened, if c1 only calculated a new shortest path if he is on the old path and the next field from c1 on that path is blocked. In that case c1 would not care about any char blocking the path as long as the other char is not directly in front of c1. It would also be more efficient this way

Yes, its a good solution, i think it should work

**Comment by Annoraaq** [comment]
I created a new issue for this:
https://github.com/Annoraaq/grid-engine/issues/104
